tests/misc-utils.c

```diff
@@ -0,0 +1,295 @@
+            return FALSE;
+
+          if (info == NULL || child == NULL)
+            break;
+
+          if (g_file_info_get_file_type (info) == G_FILE_TYPE_DIRECTORY)
+            {
+              g_queue_push_head (&queue, g_object_ref (child));
+              dir_stack = g_list_prepend (dir_stack, g_object_ref (child));
+            }
+          else if (!g_file_delete (child, NULL, error))
+            return FALSE;
+        }
+    }
+
+  while (dir_stack != NULL)
+    {
+      GList *first = dir_stack;
+      g_autoptr(GFile) dir = G_FILE (first->data);
+
+      dir_stack = g_list_remove_link (dir_stack, first);
+      g_list_free_1 (first);
+      if (!g_file_delete (dir, NULL, error))
+        return FALSE;
+    }
+
+  return TRUE;
+}
+
+gchar **
+generate_strv (const gchar *str, ...)
+{
+  g_autoptr(GPtrArray) vec = string_array_new ();
+  va_list args;
+
+  va_start (args, str);
+  while (str != NULL)
+    {
+      g_ptr_array_add (vec, g_strdup (str));
+      str = va_arg (args, const gchar *);
+    }
+  va_end (args);
+  g_ptr_array_add (vec, NULL);
+
+  return (gchar**)g_ptr_array_free (g_steal_pointer (&vec), FALSE);
+}
+
+gboolean
+load_to_bytes (GFile *file,
+               GBytes **bytes,
+               GError **error)
+{
+  g_autofree gchar *data = NULL;
+  gsize len;
+
+  if (!g_file_load_contents (file,
+                             NULL,
+                             &data,
+                             &len,
+                             NULL,
+                             error))
+    return FALSE;
+
+  *bytes = g_bytes_new_take (g_steal_pointer (&data), len);
+  return TRUE;
+}
+
+gboolean
+create_file (GFile *path,
+             GBytes* bytes,
+             GError **error)
+{
+  return g_file_replace_contents (path,
+                                  (bytes != NULL) ? g_bytes_get_data (bytes, NULL) : "",
+                                  (bytes != NULL) ? g_bytes_get_size (bytes) : 0,
+                                  NULL,
+                                  FALSE,
+                                  G_FILE_CREATE_NONE,
+                                  NULL,
+                                  NULL,
+                                  error);
+}
+
+gboolean
+create_directory (GFile *path,
+                  GError **error)
+{
+  g_autoptr(GError) local_error = NULL;
+
+  if (!g_file_make_directory_with_parents (path, NULL, &local_error))
+    if (!g_error_matches (local_error, G_IO_ERROR, G_IO_ERROR_EXISTS))
+      {
+        g_propagate_error (error, g_steal_pointer (&local_error));
+        return FALSE;
+      }
+
+  return TRUE;
+}
+
+gboolean
+create_symlink (const gchar *target,
+                GFile *link,
+                GError **error)
+{
+  g_autofree gchar *path = g_file_get_path (link);
+
+  if (symlink (target, path) < 0)
+    {
+      int saved_errno = errno;
+
+      g_set_error_literal (error, G_IO_ERROR, g_io_error_from_errno (saved_errno),
+                           g_strerror (saved_errno));
```

Comment:
is there any reason to have `saved_errno` here? I can't imagine a case where the child functions might also set `errno`. I suppose its better to have this than not to have it though...